Swapping master and slave and setting up new master/slave replication

We had a ledis master/slave setup that was replicating fine.

We recently experienced some data corruption on the master so we took it down and promoted the slave to be the new master.

Replication now no longer works between the new master and the old master (which is now the slave). The new slave had all its prior data removed and should just function like a brand new ledis slave node.

On the new master:

```
10.0.0.208:6380> info replication
# Replication
role:master
pub_log_num:0
pub_log_ack_per_time:0
slaveof:
slaves:10.0.0.99:6380
last_log_id:4102189945
first_log_id:4050241982
commit_log_id:4102189945
master_last_log_id:0
```

On the new slave:

```
10.0.0.99:6380> info replication
# Replication
role:slave
pub_log_num:0
pub_log_ack_per_time:0
slaveof:10.0.0.208:6380
last_log_id:0
first_log_id:0
commit_log_id:0
master_last_log_id:0
master_link_status:up
slave_priority:100
slave_repl_offset:0
```

Is there some additional tasks required to set up replication in this case?

When I try to start the new master with _-rpl_ it panics:

```
# /usr/local/bin/ledis-server -rpl -config=/etc/ledis/ledis.conf
panic: runtime error: index out of range

goroutine 27 [running]:
github.com/siddontang/ledisdb/store/leveldb.(*WriteBatch).Delete(0xc8201c79e0, 0xc82022e42e, 0x0, 0x106)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/store/leveldb/batch.go:58 +0x49
github.com/siddontang/ledisdb/store.(*WriteBatch).Delete(0xc8201f3d00, 0xc82022e42e, 0x0, 0x106)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/store/writebatch.go:33 +0x5c
github.com/syndtr/goleveldb/leveldb.(*Batch).Replay.func1(0x1, 0x0, 0xc82022e42e, 0x0, 0x106, 0x0, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/batch.go:129 +0x5f
github.com/syndtr/goleveldb/leveldb.(*Batch).decodeRec(0xc82026a6c0, 0xc820034d58, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/batch.go:227 +0x33f
github.com/syndtr/goleveldb/leveldb.(*Batch).Replay(0xc82026a6c0, 0x7f73bc778738, 0xc8201f3d00, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/batch.go:131 +0x53
github.com/siddontang/ledisdb/store.(*BatchData).Replay(0xc82026a6c0, 0x7f73bc778708, 0xc8201f3d00, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/store/writebatch.go:140 +0x74
github.com/siddontang/ledisdb/ledis.(*Ledis).handleReplication(0xc8201b0240, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/ledis/replication.go:58 +0x70c
github.com/siddontang/ledisdb/ledis.(*Ledis).onReplication(0xc8201b0240)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/ledis/replication.go:86 +0xf5
created by github.com/siddontang/ledisdb/ledis.Open
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/ledis/ledis.go:83 +0x46f

goroutine 1 [runnable]:
github.com/siddontang/ledisdb/ledis.(*Ledis).WaitReplication(0xc8201b0240, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/ledis/replication.go:102 +0x2b4
github.com/siddontang/ledisdb/ledis.Open(0xc8200b2780, 0xc8201a56e0, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/ledis/ledis.go:87 +0x47d
github.com/siddontang/ledisdb/server.NewApp(0xc8200b2780, 0x1d, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/server/app.go:120 +0x81a
main.main()
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/cmd/ledis-server/main.go:92 +0x4d4

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 21 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 22 [select]:
github.com/siddontang/go/log.(*Logger).run(0xc82006a150)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/Godeps/_workspace/src/github.com/siddontang/go/log/log.go:100 +0x269
created by github.com/siddontang/go/log.New
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/Godeps/_workspace/src/github.com/siddontang/go/log/log.go:80 +0x1cc

goroutine 24 [select]:
github.com/siddontang/ledisdb/server.(*snapshotStore).run(0xc8201a56e0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/server/snapshot.go:115 +0x2c3
created by github.com/siddontang/ledisdb/server.newSnapshotStore
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/server/snapshot.go:67 +0x2fb

goroutine 25 [select]:
github.com/siddontang/ledisdb/rpl.(*FileStore).checkTableReaders(0xc82006a460)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/rpl/file_store.go:285 +0x338
created by github.com/siddontang/ledisdb/rpl.NewFileStore
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/rpl/file_store.go:97 +0x298

goroutine 26 [select]:
github.com/siddontang/ledisdb/rpl.(*Replication).run(0xc82006e720)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/rpl/rpl.go:298 +0x796
created by github.com/siddontang/ledisdb/rpl.NewReplication
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/rpl/rpl.go:78 +0x5da
```

I built a new version of the server from master and I get this on the master when starting it up with *-rpl*:

```
# /var/tmp/ledis-server -rpl -config=/etc/ledis/ledis.conf
panic: runtime error: index out of range

goroutine 10 [running]:
github.com/siddontang/ledisdb/store/leveldb.(*WriteBatch).Delete(0xc420237ce0, 0xc42028042e, 0x0, 0x106)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/store/leveldb/batch.go:58 +0x57
github.com/siddontang/ledisdb/store.(*WriteBatch).Delete(0xc42024e100, 0xc42028042e, 0x0, 0x106)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/store/writebatch.go:33 +0x5c
github.com/syndtr/goleveldb/leveldb.(*Batch).Replay.func1(0x1, 0x0, 0xc42028042e, 0x0, 0x106, 0x0, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/syndtr/goleveldb/leveldb/batch.go:129 +0xb6
github.com/syndtr/goleveldb/leveldb.(*Batch).decodeRec(0xc4202c0a40, 0xc420037d60, 0x410f58, 0xa2e300)
	/home/charl/projects/go/src/github.com/syndtr/goleveldb/leveldb/batch.go:227 +0x17d
github.com/syndtr/goleveldb/leveldb.(*Batch).Replay(0xc4202c0a40, 0x7f78701da330, 0xc42024e100, 0x7f78701da330, 0xc42024e100)
	/home/charl/projects/go/src/github.com/syndtr/goleveldb/leveldb/batch.go:131 +0x67
github.com/siddontang/ledisdb/store.(*BatchData).Replay(0xc4202c0a40, 0xfb68c0, 0xc42024e100, 0xc4202c0a40, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/store/writebatch.go:140 +0x6a
github.com/siddontang/ledisdb/ledis.(*Ledis).handleReplication(0xc4201f6240, 0x0, 0x0)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/ledis/replication.go:58 +0x1b0
github.com/siddontang/ledisdb/ledis.(*Ledis).onReplication(0xc4201f6240)
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/ledis/replication.go:86 +0x102
created by github.com/siddontang/ledisdb/ledis.Open
	/home/charl/projects/go/src/github.com/siddontang/ledisdb/ledis/ledis.go:83 +0x408
```

--- a/rpl/rpl.go
+++ b/rpl/rpl.go
@@ -74,6 +74,8 @@ func NewReplication(cfg *config.Config) (*Replication, error) {
 		return nil, err
 	}
 
+	log.Infof("staring replication with commit ID %d", r.commitID)
+
 	r.wg.Add(1)
 	go r.run()
 
@@ -87,6 +89,8 @@ func (r *Replication) Close() error {
 
 	r.m.Lock()
 	defer r.m.Unlock()
+
+	log.Infof("closing replication with commit ID %d", r.commitID)
 
 	if r.s != nil {
 		r.s.Close()

--- a/store/leveldb/batch.go
+++ b/store/leveldb/batch.go
@@ -16,15 +16,12 @@ import (
 type WriteBatch struct {
 	db     *DB
 	wbatch *C.leveldb_writebatch_t
-
-	gbatch *leveldb.Batch
 }
 
 func newWriteBatch(db *DB) *WriteBatch {
 	w := new(WriteBatch)
 	w.db = db
 	w.wbatch = C.leveldb_writebatch_create()
-	w.gbatch = new(leveldb.Batch)
 
 	return w
 }
@@ -34,8 +31,6 @@ func (w *WriteBatch) Close() {
 		C.leveldb_writebatch_destroy(w.wbatch)
 		w.wbatch = nil
 	}
-
-	w.gbatch = nil
 }
 
 func (w *WriteBatch) Put(key, value []byte) {
@@ -97,9 +92,8 @@ func leveldb_writebatch_iterate_delete(p unsafe.Pointer, k *C.char, klen C.size_
 }
 
 func (w *WriteBatch) Data() []byte {
-	w.gbatch.Reset()
+	gbatch := leveldb.Batch{}
 	C.leveldb_writebatch_iterate_ext(w.wbatch,
-		unsafe.Pointer(w.gbatch))
-	b := w.gbatch.Dump()
-	return b
+		unsafe.Pointer(&gbatch))
+	return gbatch.Dump()
 }

